File: modulo2/shopapp/app/src/main/java/com/shopapp/presentation/ui/admin/orders/OrderAdminDetailScreen.kt
// presentation/ui/admin/orders/OrderAdminDetailScreen.kt
package com.shopapp.presentation.ui.admin.orders

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import com.shopapp.domain.model.Order
import com.shopapp.domain.model.OrderItem
import com.shopapp.domain.model.OrderStatus
import com.shopapp.presentation.components.LoadingScreen
import com.shopapp.presentation.components.ErrorScreen
import com.shopapp.presentation.viewmodel.OrderDetailUiState
import com.shopapp.presentation.viewmodel.OrderDetailViewModel
import com.shopapp.theme.*
import java.text.SimpleDateFormat
import java.util.Locale

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun OrderAdminDetailScreen(
    orderId:    Int,
    onBack:     () -> Unit,
    onStatusChange: (Int, OrderStatus) -> Unit,
    viewModel:  OrderDetailViewModel = hiltViewModel(),
) {
    val state by viewModel.state.collectAsState()
    LaunchedEffect(orderId) { viewModel.load(orderId) }

    when (val s = state) {
        is OrderDetailUiState.Loading ->
            LoadingScreen("Cargando pedido #$orderId...")
        is OrderDetailUiState.Error   ->
            ErrorScreen(s.message, onRetry = { viewModel.load(orderId) })
        is OrderDetailUiState.Success ->
            AdminOrderDetailContent(
                order         = s.order,
                onBack        = onBack,
                onStatusChange = { newStatus ->
                    onStatusChange(orderId, newStatus)
                    viewModel.load(orderId)
                },
            )
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun AdminOrderDetailContent(
    order:         Order,
    onBack:        () -> Unit,
    onStatusChange:(OrderStatus) -> Unit,
) {
    val inputFmt  = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSSSS'Z'", Locale.getDefault())
    val outputFmt = SimpleDateFormat("dd MMM yyyy · HH:mm", Locale("es"))
    val dateStr   = runCatching { outputFmt.format(inputFmt.parse(order.createdAt)!!) }
        .getOrDefault(order.createdAt.take(16))
    val updatedStr = runCatching { outputFmt.format(inputFmt.parse(order.updatedAt)!!) }
        .getOrDefault(order.updatedAt.take(16))

    val taxAmount = order.total - order.total / 1.15
    val subtotal  = order.total - taxAmount

    Scaffold(
        topBar = {
            TopAppBar(
                title = {
                    Column {
                        Text(
                            "Pedido #${order.id}",
                            style      = MaterialTheme.typography.titleMedium,
                            fontWeight = FontWeight.Bold,
                            color      = TextPrimary,
                        )
                        Text(
                            order.username,
                            style = MaterialTheme.typography.bodySmall,
                            color = TextSecondary,
                        )
                    }
                },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.Default.ArrowBack, null, tint = TextPrimary)
                    }
                },
                actions = {
                    // Selector de estado en la TopAppBar
                    StatusDropdown(
                        current  = order.status,
                        onChange = onStatusChange,
                        modifier = Modifier.padding(end = 12.dp),
                    )
                },
                colors = TopAppBarDefaults.topAppBarColors(containerColor = Surface),
            )
        },
        containerColor = Background,
    ) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .verticalScroll(rememberScrollState())
                .padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(14.dp),
        ) {

            // ── Info general ──────────────────────────────────
            Surface(color = Surface, shape = MaterialTheme.shapes.large, modifier = Modifier.fillMaxWidth()) {
                Column(modifier = Modifier.padding(16.dp)) {
                    SectionLabel("Información del pedido")
                    Spacer(Modifier.height(12.dp))
                    InfoGrid(listOf(
                        "Cliente"      to order.username,
                        "Fecha"        to dateStr,
                        "Actualizado"  to updatedStr,
                        "Total ítems"  to "${order.numItems}",
                    ))
                }
            }

            // ── Ítems ──────────────────────────────────────────
            Surface(color = Surface, shape = MaterialTheme.shapes.large, modifier = Modifier.fillMaxWidth()) {
                Column(modifier = Modifier.padding(16.dp)) {
                    SectionLabel("Productos (${order.numItems})")
                    Spacer(Modifier.height(12.dp))
                    Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
                        order.items.forEach { item ->
                            AdminOrderItemRow(item = item)
                        }
                    }
                }
            }

            // ── Totales ────────────────────────────────────────
            Surface(color = Surface, shape = MaterialTheme.shapes.large, modifier = Modifier.fillMaxWidth()) {
                Column(modifier = Modifier.padding(16.dp)) {
                    SectionLabel("Resumen financiero")
                    Spacer(Modifier.height(12.dp))
                    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                        FinancialRow("Subtotal (sin IVA)", subtotal, false)
                        FinancialRow("IVA (15%)",          taxAmount, false)
                        HorizontalDivider(color = Border, thickness = 0.5.dp)
                        FinancialRow("Total",              order.total, true)
                    }
                }
            }

            // ── Cambio rápido de estado ────────────────────────
            Surface(color = Surface, shape = MaterialTheme.shapes.large, modifier = Modifier.fillMaxWidth()) {
                Column(modifier = Modifier.padding(16.dp)) {
                    SectionLabel("Cambiar estado")
                    Spacer(Modifier.height(12.dp))
                    Row(
                        modifier              = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                    ) {
                        OrderStatus.entries
                            .filter { it != order.status }
                            .forEach { status ->
                                Surface(
                                    onClick   = { onStatusChange(status) },
                                    shape     = MaterialTheme.shapes.small,
                                    color     = com.shopapp.presentation.components
                                        .orderStatusColor(status).copy(alpha = 0.1f),
                                    modifier  = Modifier.weight(1f),
                                ) {
                                    Text(
                                        text     = status.label,
                                        color    = com.shopapp.presentation.components
                                            .orderStatusColor(status),
                                        fontSize = 11.sp,
                                        fontWeight = FontWeight.Bold,
                                        modifier = Modifier.padding(8.dp),
                                    )
                                }
                            }
                    }
                }
            }
        }
    }
}

// ── Sub-composables ───────────────────────────────────────────

@Composable
private fun SectionLabel(text: String) {
    Text(
        text          = text,
        style         = MaterialTheme.typography.labelSmall,
        color         = TextSecondary,
        letterSpacing = 0.8.sp,
    )
}

@Composable
private fun InfoGrid(items: List<Pair<String, String>>) {
    items.chunked(2).forEach { row ->
        Row(
            modifier              = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            row.forEach { (label, value) ->
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        text  = label,
                        style = MaterialTheme.typography.bodySmall,
                        color = TextSecondary,
                    )
                    Text(
                        text       = value,
                        style      = MaterialTheme.typography.bodyMedium,
                        fontWeight = FontWeight.SemiBold,
                        color      = TextPrimary,
                    )
                    Spacer(Modifier.height(10.dp))
                }
            }
            // Celda vacía si el row tiene 1 elemento
            if (row.size == 1) Spacer(Modifier.weight(1f))
        }
    }
}

@Composable
private fun AdminOrderItemRow(item: OrderItem) {
    Row(
        modifier              = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment     = Alignment.CenterVertically,
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier          = Modifier.weight(1f),
        ) {
            Box(
                modifier         = Modifier
                    .size(44.dp)
                    .background(Surface2, MaterialTheme.shapes.small),
                contentAlignment = Alignment.Center,
            ) {
                Text("📦", fontSize = 20.sp)
            }
            Spacer(Modifier.width(12.dp))
            Column {
                Text(
                    text       = item.productName,
                    style      = MaterialTheme.typography.bodyMedium,
                    fontWeight = FontWeight.SemiBold,
                    color      = TextPrimary,
                    maxLines   = 2,
                )
                Text(
                    text  = "$${"%.2f".format(item.unitPrice)} × ${item.quantity} ud.",
                    style = MaterialTheme.typography.bodySmall,
                    color = TextSecondary,
                )
            }
        }
        Spacer(Modifier.width(12.dp))
        Text(
            text       = "$${"%.2f".format(item.subtotal)}",
            style      = MaterialTheme.typography.bodyMedium,
            fontWeight = FontWeight.Bold,
            color      = Accent,
        )
    }
}

@Composable
private fun FinancialRow(label: String, value: Double, isFinal: Boolean) {
    Row(
        modifier              = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
    ) {
        Text(
            text       = label,
            style      = if (isFinal) MaterialTheme.typography.titleSmall
            else MaterialTheme.typography.bodyMedium,
            fontWeight = if (isFinal) FontWeight.Bold else FontWeight.Normal,
            color      = if (isFinal) TextPrimary else TextSecondary,
        )
        Text(
            text       = "$${"%.2f".format(value)}",
            style      = if (isFinal) MaterialTheme.typography.titleSmall
            else MaterialTheme.typography.bodyMedium,
            fontWeight = if (isFinal) FontWeight.ExtraBold else FontWeight.SemiBold,
            color      = if (isFinal) Accent else TextPrimary,
        )
    }
}

File: modulo4/flutter_shop_app/lib/presentation/screens/admin/order_admin_detail_screen.dart
// lib/presentation/screens/admin/order_admin_detail_screen.dart

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_shop_app/domain/model/order.dart';
import 'package:go_router/go_router.dart';
import '../../../theme/app_colors.dart';
import '../../../core/utils/formatters.dart';
import '../../widgets/status_badge.dart';
import '../../providers/orders_admin_provider.dart';
import '../../widgets/status_dropdown.dart';

class OrderAdminDetailScreen extends ConsumerWidget {
  final int orderId;
  const OrderAdminDetailScreen({super.key, required this.orderId});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final orderAsync = ref.watch(orderAdminDetailProvider(orderId));

    return orderAsync.when(
      loading: () => const Center(
        child: CircularProgressIndicator(color: AppColors.accent),
      ),
      error: (err, _) => Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Text(err.toString(), style: const TextStyle(color: AppColors.error)),
            const SizedBox(height: 12),
            ElevatedButton(
              onPressed: () => context.pop(),
              child:     const Text('Volver'),
            ),
          ],
        ),
      ),
      data: (order) => _DetailContent(order: order, ref: ref),
    );
  }
}

class _DetailContent extends StatelessWidget {
  final Order  order;
  final WidgetRef ref;
  const _DetailContent({required this.order, required this.ref});

  @override
  Widget build(BuildContext context) {
    final taxAmount  = order.total - order.total / 1.15;
    final subtotal   = order.total - taxAmount;
    final dateStr    = formatDateTime(order.createdAt);
    final updatedStr = formatDateTime(order.updatedAt);

    return SingleChildScrollView(
      padding: const EdgeInsets.all(16),
      child:   Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [

          // Info general + estado
          _Card(
            title: 'Información del pedido',
            child: Column(
              children: [
                _InfoRow('Cliente',     order.username),
                _InfoRow('Fecha',       dateStr),
                _InfoRow('Actualizado', updatedStr),
                _InfoRow('Ítems',       '${order.numItems}'),
                const SizedBox(height: 12),
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    const Text('Estado actual',
                        style: TextStyle(color: AppColors.textSecondary, fontSize: 13)),
                    StatusBadge(status: order.status),
                  ],
                ),
                const SizedBox(height: 12),
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    const Text('Cambiar estado',
                        style: TextStyle(color: AppColors.textSecondary, fontSize: 13)),
                    StatusDropdown(
                      current:  order.status,
                      onChange: (s) {
                        ref.read(ordersAdminProvider.notifier).changeStatus(order.id, s);
                        ScaffoldMessenger.of(context).showSnackBar(SnackBar(
                          content:         Text('Estado cambiado a "${s.label}"'),
                          backgroundColor: AppColors.success,
                          duration:        const Duration(seconds: 2),
                        ));
                      },
                    ),
                  ],
                ),
              ],
            ),
          ),
          const SizedBox(height: 14),

          // Ítems
          _Card(
            title: 'Productos (${order.numItems})',
            child: Column(
              children: order.items.map((item) => Padding(
                padding: const EdgeInsets.symmetric(vertical: 8),
                child:   Row(
                  children: [
                    Container(
                      width:  44, height: 44,
                      decoration: BoxDecoration(
                        color:        AppColors.surface2,
                        borderRadius: BorderRadius.circular(8),
                      ),
                      child: const Center(child: Text('📦', style: TextStyle(fontSize: 20))),
                    ),
                    const SizedBox(width: 12),
                    Expanded(
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          Text(item.productName,
                              style: const TextStyle(
                                color: AppColors.textPrimary, fontWeight: FontWeight.w600,
                              )),
                          Text('${formatPrice(item.unitPrice)} × ${item.quantity} ud.',
                              style: const TextStyle(color: AppColors.textSecondary, fontSize: 12)),
                        ],
                      ),
                    ),
                    Text(formatPrice(item.subtotal),
                        style: const TextStyle(
                          color: AppColors.accent, fontWeight: FontWeight.bold,
                        )),
                  ],
                ),
              )).toList(),
            ),
          ),
          const SizedBox(height: 14),

          // Resumen financiero
          _Card(
            title: 'Resumen financiero',
            child: Column(
              children: [
                _TotalRow('Subtotal (sin IVA)', subtotal,    false),
                const SizedBox(height: 6),
                _TotalRow('IVA (15%)',          taxAmount,   false),
                const SizedBox(height: 8),
                const Divider(),
                const SizedBox(height: 8),
                _TotalRow('Total',              order.total, true),
              ],
            ),
          ),
          const SizedBox(height: 14),

          // Cambio rápido de estado
          _Card(
            title: 'Cambio rápido de estado',
            child: Wrap(
              spacing: 8, runSpacing: 8,
              children: OrderStatus.values
                  .where((s) => s != order.status)
                  .map((s) {
                    final color = orderStatusColor(s);
                    return GestureDetector(
                      onTap: () {
                        ref.read(ordersAdminProvider.notifier).changeStatus(order.id, s);
                        ScaffoldMessenger.of(context).showSnackBar(SnackBar(
                          content:         Text('Estado cambiado a "${s.label}"'),
                          backgroundColor: AppColors.success,
                          duration:        const Duration(seconds: 2),
                        ));
                      },
                      child: Container(
                        padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 8),
                        decoration: BoxDecoration(
                          color:        color.withValues(alpha: 0.1),
                          borderRadius: BorderRadius.circular(10),
                          border:       Border.all(color: color.withValues(alpha: 0.3)),
                        ),
                        child: Row(
                          mainAxisSize: MainAxisSize.min,
                          children: [
                            Container(
                              width:  8, height: 8,
                              decoration: BoxDecoration(color: color, shape: BoxShape.circle),
                            ),
                            const SizedBox(width: 6),
                            Text(
                              s.label,
                              style: TextStyle(
                                color: color, fontWeight: FontWeight.bold, fontSize: 13,
                              ),
                            ),
                          ],
                        ),
                      ),
                    );
                  }).toList(),
            ),
          ),
          const SizedBox(height: 24),
        ],
      ),
    );
  }
}

// ── Sub-widgets ───────────────────────────────────────────────

class _Card extends StatelessWidget {
  final String title;
  final Widget child;
  const _Card({required this.title, required this.child});

  @override
  Widget build(BuildContext context) => Container(
    width:   double.infinity,
    padding: const EdgeInsets.all(16),
    decoration: BoxDecoration(
      color:        AppColors.surface,
      borderRadius: BorderRadius.circular(16),
    ),
    child: Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(
          title.toUpperCase(),
          style: const TextStyle(
            color: AppColors.textSecondary, fontSize: 11,
            fontWeight: FontWeight.bold, letterSpacing: 0.8,
          ),
        ),
        const SizedBox(height: 14),
        child,
      ],
    ),
  );
}

class _InfoRow extends StatelessWidget {
  final String label;
  final String value;
  const _InfoRow(this.label, this.value);

  @override
  Widget build(BuildContext context) => Padding(
    padding: const EdgeInsets.symmetric(vertical: 5),
    child:   Row(
      mainAxisAlignment: MainAxisAlignment.spaceBetween,
      children: [
        Text(label, style: const TextStyle(color: AppColors.textSecondary, fontSize: 13)),
        Text(value, style: const TextStyle(
          color: AppColors.textPrimary, fontWeight: FontWeight.w600, fontSize: 13,
        )),
      ],
    ),
  );
}

class _TotalRow extends StatelessWidget {
  final String label;
  final double value;
  final bool   isFinal;
  const _TotalRow(this.label, this.value, this.isFinal);

  @override
  Widget build(BuildContext context) => Row(
    mainAxisAlignment: MainAxisAlignment.spaceBetween,
    children: [
      Text(
        label,
        style: TextStyle(
          color:      isFinal ? AppColors.textPrimary : AppColors.textSecondary,
          fontSize:   isFinal ? 16 : 14,
          fontWeight: isFinal ? FontWeight.bold : FontWeight.normal,
        ),
      ),
      Text(
        formatPrice(value),
        style: TextStyle(
          color:      isFinal ? AppColors.accent : AppColors.textPrimary,
          fontSize:   isFinal ? 18 : 14,
          fontWeight: isFinal ? FontWeight.w800 : FontWeight.w600,
        ),
      ),
    ],
  );
}